### PR TITLE
kaggle cifar-10, dogs Try all gpus

### PR DESCRIPTION
**Pull request by xiaotinghe**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_computer-vision/kaggle-dog.md [hunk 1]
```diff
@@ -273,6 +288,6 @@ method for submitting results is similar to method in
 1. Do you get better results if you use a deeper pre-trained model?
 1. Scan the QR code to access the relevant discussions and exchange ideas about the methods used and the results obtained with the community. Can you come up with any better techniques?
 
-:begin_tab:`mxnet`
-[Discussions](https://discuss.d2l.ai/t/380)
-:end_tab:
+## [Discussions](https://discuss.mxnet.io/t/2451)
+
+![](../img/qr_kaggle-gluon-dog.svg)
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-dog.md:293]: don't change this

File: chapter_computer-vision/kaggle-cifar10.md [hunk 2]
```diff
@@ -351,6 +356,6 @@ for submitting results is similar to method in :numref:`sec_kaggle_house`.
 1. What accuracy can you achieve when not using image augmentation?
 1. Scan the QR code to access the relevant discussions and exchange ideas about the methods used and the results obtained with the community. Can you come up with any better techniques?
 
-:begin_tab:`mxnet`
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-cifar10.md:354]: don't change this

File: chapter_computer-vision/kaggle-cifar10.md [hunk 3]
```diff
@@ -141,10 +141,10 @@ def reorg_cifar10_data(data_dir, valid_ratio):
     reorg_test(data_dir)
 ```
 
-We only set the batch size to $1$ for the demo dataset. During actual training and testing, the complete dataset of the Kaggle competition should be used and `batch_size` should be set to a larger integer, such as $128$. We use $10\%$ of the training examples as the validation set for tuning hyperparameters.
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-cifar10.md:144]: keep all  "hyperparameters"

File: chapter_computer-vision/kaggle-cifar10.md [hunk 4]
```diff
@@ -270,61 +270,66 @@ loss = gluon.loss.SoftmaxCrossEntropyLoss()
...
         if valid_iter is not None:
-            valid_acc = d2l.evaluate_accuracy_gpu(net, valid_iter)
-            epoch_s = (f'epoch {epoch + 1}, loss {train_l_sum / n:f}, '
-                       f'train acc {train_acc_sum / n:f}, '
-                       f'valid acc {valid_acc:f}, ')
-        else:
-            epoch_s = (f'epoch {epoch + 1}, loss {train_l_sum / n:f}, '
-                       f'train acc {train_acc_sum / n:f}, ')
-        print(epoch_s + time_s + ', lr ' + str(trainer.learning_rate))
+            valid_acc = d2l.evaluate_accuracy_gpus(net, valid_iter, d2l.split_batch)
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-cifar10.md:299]: don't change f-string

File: chapter_computer-vision/kaggle-cifar10.md [hunk 5]
```diff
@@ -141,10 +141,10 @@ def reorg_cifar10_data(data_dir, valid_ratio):
     reorg_test(data_dir)
 ```
 
-We only set the batch size to $1$ for the demo dataset. During actual training and testing, the complete dataset of the Kaggle competition should be used and `batch_size` should be set to a larger integer, such as $128$. We use $10\%$ of the training examples as the validation set for tuning hyperparameters.
+We only set the batch size to $1$ for the demo dataset. During actual training and testing, the complete dataset of the Kaggle competition should be used and `batch_size` should be set to a larger integer, such as $128$. We use $10\%$ of the training examples as the validation set for tuning hyper-parameters.
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-cifar10.md:144]: change in text: 1 -> 4

File: chapter_computer-vision/kaggle-cifar10.md [hunk 6]
```diff
@@ -270,61 +270,66 @@ loss = gluon.loss.SoftmaxCrossEntropyLoss()
...
 ## Training and Validating the Model
 
-Now, we can train and validate the model. The following hyperparameters can be tuned. For example, we can increase the number of epochs. Because `lr_period` and `lr_decay` are set to 80 and 0.1 respectively, the learning rate of the optimization algorithm will be multiplied by 0.1 after every 80 epochs. For simplicity, we only train one epoch here.
+Now, we can train and validate the model. The following hyper-parameters can be tuned. For example, we can increase the number of epochs. Because `lr_period` and `lr_decay` are set to 80 and 0.1 respectively, the learning rate of the optimization algorithm will be multiplied by 0.1 after every 80 epochs. For simplicity, we only train one epoch here.
 
 ```{.python .input  n=13}
-ctx, num_epochs, lr, wd = d2l.try_gpu(), 1, 0.1, 5e-4
-lr_period, lr_decay, net = 80, 0.1, get_net(ctx)
+ctx, num_epochs, lr, wd = d2l.try_all_gpus(), 5, 0.1, 5e-4
+lr_period, lr_decay, net = 50, 0.1, get_net(ctx)
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-cifar10.md:317]: change to 80, then re-test on the full dataset

File: chapter_computer-vision/kaggle-dog.md [hunk 7]
```diff
@@ -172,60 +173,73 @@ def get_net(ctx):
...
+                animator.add(epoch + i / num_batches, 
+                             (metric[0] / metric[1], None))
         if valid_iter is not None:
-            valid_loss = evaluate_loss(valid_iter, net, ctx)
-            epoch_s = (f'epoch {epoch + 1}, train loss {train_l_sum / n:f}, '
-                       f'valid loss {valid_loss:f}, ')
-        else:
-            epoch_s = f'epoch {epoch + 1}, train loss {train_l_sum / n:f}, '
-        print(epoch_s + time_s + ', lr ' + str(trainer.learning_rate))
+            valid_loss = evaluate_loss(valid_iter, net, ctx_list)
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-dog.md:226]: don;t change this, use f-string

File: chapter_computer-vision/kaggle-dog.md [hunk 8]
```diff
@@ -234,14 +248,15 @@ train(net, train_iter, valid_iter, num_epochs, lr, wd, ctx, lr_period,
...
+After obtaining a satisfactory model design and hyper-parameters, we use all training datasets (including validation sets) to retrain the model and then classify the testing set. Note that predictions are made by the output network we just trained.
 
-```{.python .input  n=8}
+```{.python .input  n=13}
 net = get_net(ctx)
 net.hybridize()
 train(net, train_valid_iter, None, num_epochs, lr, wd, ctx, lr_period,
       lr_decay)
 
+ctx=d2l.try_gpu()
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-dog.md:259]: add space around operator =

File: chapter_computer-vision/kaggle-dog.md [hunk 9]
```diff
@@ -172,60 +173,73 @@ def get_net(ctx):
...
-        train_l_sum, n, start = 0.0, 0, time.time()
+        metric = d2l.Accumulator(2)
         if epoch > 0 and epoch % lr_period == 0:
             trainer.set_learning_rate(trainer.learning_rate * lr_decay)
-        for X, y in train_iter:
-            y = y.as_in_ctx(ctx)
-            output_features = net.features(X.as_in_ctx(ctx))
+        for i, (features, labels) in enumerate(train_iter):
+            timer.start()
+            X_shards, y_shards = d2l.split_batch(features, labels, ctx_list)
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-dog.md:211]: add a flag to d2l.split_batch(): even_split=True
Here pass in `False`

File: chapter_computer-vision/kaggle-dog.md [hunk 10]
```diff
@@ -125,7 +126,7 @@ train_ds, valid_ds, train_valid_ds, test_ds = [
 
 Here, we create `DataLoader` instances, just like in :numref:`sec_kaggle_cifar10`.
 
-```{.python .input}
+```{.python .input  n=7}
 train_iter, train_valid_iter = [gluon.data.DataLoader(
     dataset.transform_first(transform_train), batch_size, shuffle=True, 
     last_batch='keep') for dataset in (train_ds, train_valid_ds)]
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-dog.md:132]: rollover

File: chapter_computer-vision/kaggle-dog.md [hunk 11]
```diff
@@ -172,60 +173,73 @@ def get_net(ctx):
...
+def evaluate_loss(data_iter, net, ctx_list):
     l_sum, n = 0.0, 0
-    for X, y in data_iter:
-        y = y.as_in_ctx(ctx)
-        output_features = net.features(X.as_in_ctx(ctx))
-        outputs = net.output_new(output_features)
-        l_sum += float(loss(outputs, y).sum())
-        n += y.size
+    for features, labels in data_iter:
+        X_shards, y_shards = d2l.split_batch(features, labels, ctx_list)
```
Comment (astonzhang) [inline chapter_computer-vision/kaggle-dog.md:182]: add `even_split=True` in d2l.split_batch
Here pass in `False`

**Comment by astonzhang** [comment]
Besides, for multi-gpu pls re-test:

train_iter: discard
train_valid_iter: discard
valid_iter: discard
test_iter: keep

**Comment by mli** [comment]
Job d2l-en/PR-1128/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1128/

**Comment by mli** [comment]
Job d2l-en/PR-1128/3 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1128/

**Comment by astonzhang** [comment]
1. Increase cifar-10 test acc to > 92.5. Let me know running time for 300 epochs
2. My todo: kaggle_dog_tiny_v2 -> kaggle_dog_tiny

**Comment by xiaotinghe** [comment]
Running time about 5000s

**Comment by xiaotinghe** [comment]
test acc 93.03

**Comment by mli** [comment]
Job d2l-en/PR-1128/5 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1128/

**Comment by astonzhang** [comment]
> test acc 93.03

Is it due to the change of lr_period from 80 to 50?
